### PR TITLE
WIP: Allow junctions to inject dependencies from current project to junctioned one

### PR DESCRIPTION
**Pull request by Cynical-Optimist**
[See original merge request on GitLab](https://gitlab.com/BuildStream/buildstream/-/merge_requests/1913)
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 10, 2020, 16:13

There are cases where downstream projects want to overwrite upstream
elements.

So far, either the downstream project just allows overlaps and hopes
that ABI is compatible. There are also files left from original
elements when not overwritten.

Or downstream adds an "overlay" with local source to the
junctions. However on the latter, elements in the overlay are
considered as part of upstream, which means in cannot depend on
element outside of upstream (unless cyclic junction maybe?).

Here we add a feature to junctions to allow injecting elements from
current project to junctions.

This is useful for example in the case of GNOME SDK needing to override
GLib, GObjectIntrospection, libsoup, etc. from Freedesktop SDK.

The backported branch to 1.4 !1914 was tested on GNOME SDK: https://gitlab.gnome.org/GNOME/gnome-build-meta/-/merge_requests/658

### CODE REVIEW

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 10, 2020, 16:13

marked as a **Work In Progress**

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 10, 2020, 16:13

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 10, 2020, 16:14

I will test it for real with GNOME SDK before unmarking it as WIP.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on May 10, 2020, 18:37

Depending on how this is used, this can cause (silent) problems. E.g., if a downstream project uses this to apply an extra patch to an element, the downstream project will miss all future element changes including security updates and other critical bug fixes from the upstream BuildStream project (unless they actively track that somehow).

This is probably not a concern for the GNOME SDK use case where the replaced elements are part of the GNOME project. However, if BuildStream were to encourage this workflow, I would guess that it would be used in cases where the above update issue is a real concern.

Any thoughts on how we could mitigate this issue?

Also, this is related to !1901, which introduces overrides for junction elements.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 10, 2020, 21:11

added 1 commit

<ul><li>7b8483bc - Allow junctions to inject dependencies from current project to junctioned one</li></ul>

[Compare with previous version](/BuildStream/buildstream/-/merge_requests/1913/diffs?diff_id=89734863&start_sha=a99b1b993d1df9bb8117220d3f0eadea111283f6)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 10, 2020, 21:16

[[Gitlab user @juergbi]](https://gitlab.com/juergbi), maybe we could have some assertions. Like the hash of the file.

However, there is the opposite issue. Downstream fixes a security issue which upstream has not. If it is a static library, or if the fix requires changes in header (C++ inline method), or pkgconfig configuration, then not having this feature introduces a security problem.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 10, 2020, 21:18

I did not follow the mailing list. So I did not see someone else tried to fix this. Sorry for that.

I have now tested it on GNOME SDK, and it can build it.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on May 10, 2020, 21:38

> However, there is the opposite issue. Downstream fixes a security issue which upstream has not. If it is a static library, or if the fix requires changes in header (C++ inline method), or pkgconfig configuration, then not having this feature introduces a security problem.

The best option in these cases is to submit the security fix upstream, of course, otherwise other downstreams still have the security problem. If upstream can't apply the fix in time, downstream can always create their own branch of the upstream project with the fix in place. That way git can be used to merge/rebase changes, avoiding the issue I've described.

In my opinion, using a git branch for security fixes and similar cases is a reasonable approach (only needed if upstream is slow/uncooperative or if there is a disagreement in configuration/versions/patches). However, I suspect it might not be very convenient for the GNOME use case, and I suspect the same for the hash assertion.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 11, 2020, 07:31

First to clarify, while !1901 may be related, it is really a different thing.


This situation comes up a lot, and I’m really not fond of the approaches to the solution which involve modifications of a junctioned project, as changing the content of a junctioned project invalidates any guarantees that the upstream may make, and also leads to friction upon upgrades to the upstream.

I think that instead we can have an approach where a downstream can consume output of an upstream more selectively, excluding entire elements from an upstream while keeping the upstream project in tact.

For instance, we could add a feature to the `compose` element which allows selection of elements to include and exclude from a dependency graph... this would allow a downstream to more cleanly replace `glib.bst` from a junctioned dependency stack.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert) on May 11, 2020, 08:07

I am not fond of such a feature on junctions, and think we should discuss this on the ML before landing it.

However, could that not already be solved today with a quilt patch applied to the junction?

I often end up doing:

```yml
kind: junction

sources:
- kind: git
  url: my_junctionned_project
- kind: quilt
  path: fixes_to_junction
```

To apply small targeted changes when I want to override the junction.

I think that this would solve your problem?

Edit: I hadn't seen the restriction on dependencies. So the quilt patch would probably not work. What is preventing you from also push to the junction the other dependencies?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 11, 2020, 10:04

[[Gitlab user @BenjaminSchubert]](https://gitlab.com/BenjaminSchubert), We do not want to fix a junction. Because then we have to move all the dependencies to the junction as well. So it is as practical as forking upstream.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 11, 2020, 10:11

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb), I do not see how using compose would work here. In my experience, compose works only at then end of the graph. Within the graph it is really painful since it adds up quite lot of unintended overlaps. So it is very difficult to work with.

Unless we make a new kind of plugin, which is a bit like compose, excepts it automatically cancels other elements when it is taken in a runtime dependency graph. A bit like a "provide" of package managements.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 11, 2020, 10:16

added 1 commit

<ul><li>0d081b63 - Allow junctions to inject dependencies from current project to junctioned one</li></ul>

[Compare with previous version](/BuildStream/buildstream/-/merge_requests/1913/diffs?diff_id=89799174&start_sha=7b8483bcd8db8cfbdaf0e53916d3fbcff4238352)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on May 11, 2020, 10:38

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jjardon]](https://gitlab.com/jjardon) on May 11, 2020, 13:56

mentioned in merge request !1914

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 13, 2020, 14:16

> Within the graph it is really painful since it adds up quite lot of unintended overlaps. So it is very difficult to work with.

If it is very difficult to work with, I think this in itself is worth solving; we should make our basics more usable.

This is getting a bit tangential and probably not the right place to discuss this, but I was thinking the other night that it might be interesting to give `compose` another feature; to optionally distill the `split rules` of it's dependencies in the resulting `compose` artifact public data.

This could be an opt-in feature, and I think this could work like:

* First we would gather the actual file lists for each element included in the compose element, and for each split domain which ends up being included
* Then we would compose new `split-rules` and record that in the public data
  * These composed split rules would not contain any wildcards and would have all variables expanded, the point would be to not conflict with anything, but just preserve the splitting context of included elements

The goal of this would be to make the composition of a selection of elements more transparent, without losing the context of split rules declared on individual elements included in the composition, I think this would make things a bit easier to work with when building on top of a composition.

(I know that in the `freedesktop-sdk` project we intentionally rewrite split-rules, I think this is besides the point, though).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Sep 28, 2020, 07:27

Months later turning my attention back to this issue... will follow up on the mailing list to.

For this thread, I'll just clarify that my intention with a `compose` element feature was to allow filtering out elements in addition to it's current ability to filter out split domains, this would **absolutely not** achieve the same as this proposed patch does, but it should make it easier for a downstream project to build and replace elements which are *gaps* in the upstream project, to relink elements at runtime in a composition which is lacking the elements we rebuild.

In this sense:

* `gnome-build-meta` would maintain a composition (`compose` element) which it would use at the base of it's builds
* This compose element would not filter out any split domains, instead it would filter out `glib.bst` and `gstreamer` components, which GNOME wants to build for itself
* Then GNOME would proceed to rebuild `glib` and `gstreamer` at the versions and with the configuration switches it likes
  * Reverse dependencies of `glib` and `gstreamer` found in the `freedesktop-sdk` composition, would only be dynamically relinked using the replaced libraries.

In other words, it seems to me to be a more straight forward way of doing what we currently do with the `filter` elements.

That said, there is some interesting merit in causing the junctioned project to rebuild *its* dependencies against an element declaration provided from a downstream, let's follow up on [Abderrahim's post](https://mail.gnome.org/archives/buildstream-list/2020-May/msg00013.html) again and see.
